### PR TITLE
add recurrence to google calendar

### PR DESCRIPTION
**Pull request by nticaric**
## Recurrence for google calendar

This PR adds recurrence to google calendar. It uses a `recurrence` parameter with a standard RRULE.  Its used like this:

```html
<div class="atcb" style="display:none;">
  {
    "name":"Add the title of your event",
    "description":"A nice description does not hurt",
    "startDate":"2022-02-21",
    "endDate":"2022-03-24",
    "startTime":"10:13",
    "endTime":"17:57",
    "location":"Somewhere over the rainbow",
    "recurrence": "RRULE:FREQ=WEEKLY;UNTIL=20110701T170000Z"
    "label":"Add to Calendar",
    "options":[
      "Google"
    ],
    "timeZone":"Europe/Berlin",
    "timeZoneOffset":"+01:00",
    "trigger":"click",
    "iCalFileName":"Reminder-Event"
  }
</div>
```

### CODE REVIEW

File: assets/js/atcb.js [hunk 1]
```diff
@@ -542,6 +542,10 @@ function atcb_generate_google(data) {
   if (data['description'] != null && data['description'] != '') {
     url += '&details=' + encodeURIComponent(data['description']);
   }
+  if (data['recurrence'] != null && data['recurrence'] != '') {
+    url += '&recur=' + encodeURIComponent(data['recurrence']);
```
Comment (jekuer) [inline assets/js/atcb.js:546]: Codacy has a fix for the issue: [Replace `'&recur='·+·encodeURIComponent(data['recurrence'` with `"&recur="·+·encodeURIComponent(data["recurrence"`](https://app.codacy.com/gh/jekuer/add-to-calendar-button/pullRequest?prid=9703945)
```suggestion
    url += "&recur=" + encodeURIComponent(data["recurrence"]);
```

File: assets/js/atcb.js [hunk 2]
```diff
@@ -542,6 +542,10 @@ function atcb_generate_google(data) {
   if (data['description'] != null && data['description'] != '') {
     url += '&details=' + encodeURIComponent(data['description']);
   }
+  if (data['recurrence'] != null && data['recurrence'] != '') {
```
Comment (jekuer) [inline assets/js/atcb.js:545]: Codacy has a fix for the issue: [Replace `'recurrence']·!=·null·&&·data['recurrence']·!=·''` with `"recurrence"]·!=·null·&&·data["recurrence"]·!=·""`](https://app.codacy.com/gh/jekuer/add-to-calendar-button/pullRequest?prid=9703945)
```suggestion
  if (data["recurrence"] != null && data["recurrence"] != "") {
```

File: assets/js/atcb.js (hunk 1 above)
Comment (jekuer) [inline assets/js/atcb.js:546]: Codacy has a fix for the issue: [Replace `'&recur='·+·encodeURIComponent(data['recurrence'` with `"&recur="·+·encodeURIComponent(data["recurrence"`](https://app.codacy.com/gh/jekuer/add-to-calendar-button/pullRequest?prid=9703945)
```suggestion
    url += "&recur=" + encodeURIComponent(data["recurrence"]);
```

File: assets/js/atcb.js (hunk 2 above)
Comment (jekuer) [inline assets/js/atcb.js:545]: Codacy has a fix for the issue: [Replace `'recurrence']·!=·null·&&·data['recurrence']·!=·''` with `"recurrence"]·!=·null·&&·data["recurrence"]·!=·""`](https://app.codacy.com/gh/jekuer/add-to-calendar-button/pullRequest?prid=9703945)
```suggestion
  if (data["recurrence"] != null && data["recurrence"] != "") {
```

File: assets/js/atcb.js (hunk 1 above)
Comment (joshsmith) [reply assets/js/atcb.js:546]: Are these Codacy fixes ignoring project preferences for single quotes?

Comment (jekuer) [reply assets/js/atcb.js:546]: Actually, there were no explicit project preferences.
I started to add respective linting rules to keep this clean.
Codacy ran on the main branch by accident.
I just merged the clean-up branch (v1.9.0), which now holds all the stuff to keep code style clean and stable.

File: assets/js/atcb.js [hunk 3]
```diff
@@ -654,6 +661,9 @@ function atcb_generate_ical(data) {
    }
    if (data['location'] != null && data['location'] != '') {
      ics_lines.push("LOCATION:" + data['location']);
+   }   
+   if (data['recurrence'] != null && data['recurrence'] != '') {
+     ics_lines.push(data['recurrence']);
```
Comment (jekuer) [inline assets/js/atcb.js:666]: Codacy has a fix for the issue: [Replace `·ics_lines.push(data['recurrence'` with `ics_lines.push(data["recurrence"`](https://app.codacy.com/gh/jekuer/add-to-calendar-button/pullRequest?prid=9703945)
```suggestion
    ics_lines.push(data["recurrence"]);
```

File: assets/js/atcb.js [hunk 4]
```diff
@@ -566,6 +570,9 @@ function atcb_generate_yahoo(data) {
   }
   if (data['description'] != null && data['description'] != '') {
     url += '&desc=' + encodeURIComponent(data['description']);
+  } 
+  if (data['recurrence'] != null && data['recurrence'] != '') {
+    url += '&recur=' + encodeURIComponent(data['recurrence']);
```
Comment (jekuer) [inline assets/js/atcb.js:575]: Codacy has a fix for the issue: [Replace `'&recur='·+·encodeURIComponent(data['recurrence'` with `"&recur="·+·encodeURIComponent(data["recurrence"`](https://app.codacy.com/gh/jekuer/add-to-calendar-button/pullRequest?prid=9703945)
```suggestion
    url += "&recur=" + encodeURIComponent(data["recurrence"]);
```

File: assets/js/atcb.js [hunk 5]
```diff
@@ -566,6 +570,9 @@ function atcb_generate_yahoo(data) {
   }
   if (data['description'] != null && data['description'] != '') {
     url += '&desc=' + encodeURIComponent(data['description']);
+  } 
+  if (data['recurrence'] != null && data['recurrence'] != '') {
```
Comment (jekuer) [inline assets/js/atcb.js:574]: Codacy has a fix for the issue: [Replace `'recurrence']·!=·null·&&·data['recurrence']·!=·''` with `"recurrence"]·!=·null·&&·data["recurrence"]·!=·""`](https://app.codacy.com/gh/jekuer/add-to-calendar-button/pullRequest?prid=9703945)
```suggestion
  if (data["recurrence"] != null && data["recurrence"] != "") {
```

**Comment by jekuer** [comment]
@nticaric , Thanks for the MR!
Unfortunately, I was changing a lot of code in the meantime, which results in some conflicts now. Will need to look into it later.

Regarding the feature:
- we should aim for some standard config style. Therefore, I would suggest to provide a clear enum list for "recurrence" (daily, weekly, monthly, yearly, ...) and add "endRecurrence" as additional option with the same date format as with "endDate".
- we should investigate, what would be a smart approach to deal with things like "every 2nd Wednesday".
- last, but not least, we should only merge it to the main branch, when we have all calendars covered. If during research, we find out that some do not support it, we should add a logic to not include them (even if specified by the user), if at the same time, "recurrence" is set. Otherwise, it is hardly usable by the mass of people.

What would be your recommendation to proceed?
We could start some kind of dev branch to merge this in for the moment.

**Comment by nticaric** [comment]
Thanks for the comment. Let me try to explain my vision of this.
The `RRULE` is defined by specification and can cover all the cases, like every 2nd Monday in a month, duration, end date, counts etc. Basically anything you can think of. So it would be impossible for us to create enums. It's more like regex for recurring events. 

In the documentation, we can point to the [specs](https://www.rfc-editor.org/rfc/rfc5545#section-3.3.10), so that the user knows what a `RRULE` is, and maybe to a `RRULE` generator like [this one](https://jakubroztocil.github.io/rrule/)

Regarding the config, I would leave it as it is now. 

```json
"recurrence": "RRULE:A_VALID_RRULE_GOES_HERE"
```

We add support to all calendars that support it and for others not.

**Pull request merged**